README.md
# express-endpoints

[![TypeScript](https://img.shields.io/badge/--3178C6?logo=typescript&logoColor=ffffff)](https://www.typescriptlang.org/)
[![JavaScript](https://img.shields.io/badge/--F7DF1E?logo=javascript&logoColor=000)](https://www.javascript.com/)
[![Visual Studio Code](https://img.shields.io/badge/--007ACC?logo=visual%20studio%20code&logoColor=ffffff)](https://code.visualstudio.com/)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-success.svg)](http://commitizen.github.io/cz-cli/)
[![semantic-release: angular](https://img.shields.io/badge/semantic--release-angular-e10079?logo=semantic-release)](https://github.com/semantic-release/semantic-release)
![node-current](https://img.shields.io/node/v/@igwtcode/express-endpoints?color=green)
[![GitHub](https://badgen.net/badge/icon/github?icon=github&label=@igwtcode/express-endpoints)](https://github.com/igwtcode/express-endpoints)
[![GitHub license](https://img.shields.io/github/license/igwtcode/express-endpoints?color=informational)](https://github.com/igwtcode/express-endpoints/blob/main/LICENSE)
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/igwtcode/express-endpoints?color=informational&logo=github)](https://github.com/igwtcode/express-endpoints/releases)
[![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/igwtcode/express-endpoints?color=informational&logo=github)](https://github.com/igwtcode/express-endpoints/tags)
[![npm](https://img.shields.io/npm/v/@igwtcode/express-endpoints?color=informational&logo=npm)](https://www.npmjs.com/package/@igwtcode/express-endpoints)
[![npm](https://img.shields.io/npm/dt/@igwtcode/express-endpoints?color=informational&logo=npm)](https://www.npmjs.com/package/@igwtcode/express-endpoints)
[![main](https://github.com/igwtcode/express-endpoints/actions/workflows/main.yml/badge.svg)](https://github.com/igwtcode/express-endpoints/actions/workflows/main.yml)
[![release](https://github.com/igwtcode/express-endpoints/actions/workflows/release.yml/badge.svg)](https://github.com/igwtcode/express-endpoints/actions/workflows/release.yml)

easy way to extract and print list of express app rest api endpoints

## Installation

```bash
npm install @igwtcode/express-endpoints
```

<details>
<summary><small>install express</small></summary>

```bash
npm install express
npm install -D @types/express # for typescript
```

</details>

## Usage

### javascript

#### import

##### module

<details>
<summary><code>package.json</code></summary>

```json
{
  "type": "module",
  "scripts": {
    "start": "node app.js"
  },
  "dependencies": {
    "express": "^4.18.2",
    "@igwtcode/express-endpoints": "^1.0.0"
  }
}
```

</details>

```javascript
import express, { Router } from 'express';
import { ExpressEndpoints } from '@igwtcode/express-endpoints';

const app = express();
const router = Router();
```

##### commonjs

<details>
<summary><code>package.json</code></summary>

```json
{
  "type": "commonjs",
  "scripts": {
    "start": "node app.js"
  },
  "dependencies": {
    "express": "^4.18.2",
    "@igwtcode/express-endpoints": "^1.0.0"
  }
}
```

</details>

```javascript
const express = require('express');
const { ExpressEndpoints } = require('@igwtcode/express-endpoints');

const app = express();
const router = express.Router();
```

#### example

```javascript
router.get('/single', (req, res) => res.send(`${req.method} ${req.path}`));
router.post('', (req, res) => res.send(`${req.method} ${req.path}`));
router.put('/:id', (req, res) => res.send(`${req.method} ${req.path} ${req.params.id}`));
router.delete('/:id', (req, res) =>
  res.send(`${req.method} ${req.path} ${req.params.id}`),
);

app.use('/tags', router);

app.get('/testGet', (req, res) => res.send(`${req.method} ${req.path}`));
app.post('/testPost', (req, res) => res.send(`${req.method} ${req.path}`));

app.use((err, req, res, next) => res.status(500).send('error'));

app.all('*', (req, res, next) => res.sendStatus(404));

const endpoints = new ExpressEndpoints(app);

console.log('*** with default options');
endpoints.print();

console.log('\n*** with color and custom prefix');
endpoints.print({ color: true, prefix: 'My-Api' });

console.log('\n*** colorized and short');
endpoints.print({ color: true, short: true });

console.log('\n*** items list', endpoints.items);

app.listen(3000, () => console.log('\nlistening on port 3000'));
```

### typescript

#### import

##### module

<details>
<summary><code>package.json</code></summary>

```json
{
  "type": "module",
  "scripts": {
    "start": "ts-node-esm app.ts"
  },
  "dependencies": {
    "express": "^4.18.2",
    "@igwtcode/express-endpoints": "^1.0.0"
  },
  "devDependencies": {
    "@types/express": "^4.17.16",
    "@types/node": "^18.11.18",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.5"
  }
}
```

</details>

<details>
<summary><code>tsconfig.json</code></summary>

```json
{
  "compilerOptions": {
    "target": "ES6",
    "module": "Node16",
    "moduleResolution": "node16",
    "esModuleInterop": true
  }
}
```

</details>

##### commonjs

<details>
<summary><code>package.json</code></summary>

```json
{
  "type": "commonjs",
  "scripts": {
    "start": "ts-node app.ts"
  },
  "dependencies": {
    "express": "^4.18.2",
    "@igwtcode/express-endpoints": "^1.0.0"
  },
  "devDependencies": {
    "@types/express": "^4.17.16",
    "@types/node": "^18.11.18",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.5"
  }
}
```

</details>

<details>
<summary><code>tsconfig.json</code></summary>

```json
{
  "compilerOptions": {
    "target": "ES6",
    "module": "CommonJS",
    "moduleResolution": "node",
    "esModuleInterop": true
  }
}
```

</details>

```typescript
import express, { Request, Response, Router, NextFunction } from 'express';
import { ExpressEndpoints } from '@igwtcode/express-endpoints';

const app = express();
const router = Router();
```

#### example

```typescript
router.get('/single', (req: Request, res: Response) =>
  res.send(`${req.method} ${req.path}`),
);
router.post('', (req: Request, res: Response) => res.send(`${req.method} ${req.path}`));
router.put('/:id', (req: Request, res: Response) =>
  res.send(`${req.method} ${req.path} ${req.params.id}`),
);
router.delete('/:id', (req: Request, res: Response) =>
  res.send(`${req.method} ${req.path} ${req.params.id}`),
);

app.use('/tags', router);

app.get('/testGet', (req: Request, res: Response) =>
  res.send(`${req.method} ${req.path}`),
);
app.post('/testPost', (req: Request, res: Response) =>
  res.send(`${req.method} ${req.path}`),
);

app.use((err: any, req: Request, res: Response, next: NextFunction) =>
  res.status(500).send('error'),
);

app.all('*', (req: Request, res: Response, next: NextFunction) => res.sendStatus(404));

const endpoints = new ExpressEndpoints(app);

console.log('*** with default options');
endpoints.print();

console.log('\n*** with color and custom prefix');
endpoints.print({ color: true, prefix: 'My-Api' });

console.log('\n*** colorized and short');
endpoints.print({ color: true, short: true });

console.log('\n*** items list', endpoints.items);

app.listen(3000, () => console.log('\nlistening on port 3000'));
```

## Sample Output

![output-snapshot-01](https://github.com/igwtcode/express-endpoints/blob/main/docs/output-snapshot-01.png)

## License

MIT ([see license](https://github.com/igwtcode/express-endpoints/blob/main/LICENSE))


docs/CHANGELOG.md
## [1.0.4](https://github.com/igwtcode/express-endpoints/compare/v1.0.3...v1.0.4) (2023-03-25)


### Bug Fixes

* **deps:** update package.json ([973b2e1](https://github.com/igwtcode/express-endpoints/commit/973b2e1681785d7dcd99331da39ab3a6d13d02ab))

## [1.0.3](https://github.com/igwtcode/express-endpoints/compare/v1.0.2...v1.0.3) (2023-03-09)


### Bug Fixes

* **deps:** update package.json ([3f5f04f](https://github.com/igwtcode/express-endpoints/commit/3f5f04f2d39c26c3af52551792dbd738b8697822)), closes [#11](https://github.com/igwtcode/express-endpoints/issues/11) [#12](https://github.com/igwtcode/express-endpoints/issues/12) [#13](https://github.com/igwtcode/express-endpoints/issues/13) [#14](https://github.com/igwtcode/express-endpoints/issues/14) [#15](https://github.com/igwtcode/express-endpoints/issues/15)

## [1.0.2](https://github.com/igwtcode/express-endpoints/compare/v1.0.1...v1.0.2) (2023-02-12)

## [1.0.1](https://github.com/igwtcode/express-endpoints/compare/v1.0.0...v1.0.1) (2023-02-07)


### Bug Fixes

* **readme:** add package name to examples, remove extra keywords, modify badges ([ae4436f](https://github.com/igwtcode/express-endpoints/commit/ae4436fee1f095fb8b5cce31236d46212a795636))

# 1.0.0 (2023-02-07)


src/test/express-endpoint.test.ts
import { ExpressEndpoints } from '@lib/express-endpoint.js';
import { Endpoint } from '@lib/interface.js';
import express, { Router } from 'express';

describe('given app has no routes', () => {
  const app = express();
  const endpoints = new ExpressEndpoints(app);

  it('endpoints should be defined and have correct type', () => {
    expect(endpoints).not.toBeUndefined();
    expect(endpoints).not.toBeNull();
    expect(endpoints).toBeInstanceOf(ExpressEndpoints);
    expect(endpoints).toEqual(expect.any(ExpressEndpoints));
  });

  it('items array should exist and have correct type', () => {
    expect(endpoints).toHaveProperty('items');
    expect(endpoints.items).not.toBeUndefined();
    expect(endpoints.items).not.toBeNull();
    expect(Array.isArray(endpoints.items)).toBeTruthy();
  });

  it('print function should exist, have correct type', () => {
    expect(endpoints).toHaveProperty('print');
    expect(endpoints.print).not.toBeUndefined();
    expect(endpoints.print).not.toBeNull();
    expect(endpoints.print).toEqual(expect.any(Function));
  });

  it('endpoint items should return empty list', () => {
    expect(endpoints.items).toHaveLength(0);
  });
});

describe('given app has 1 middleware route', () => {
  const expectedEndpoint0: Endpoint = {
    method: 'GET',
    type: 'middleware',
    route: '/testMiddlewareRoute',
  };
  const app = express();
  const mockRouteFn = jest.fn();
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const logSpy = jest.spyOn(global.console, 'log').mockImplementation(() => {});

  app.get(expectedEndpoint0.route, mockRouteFn);
  const endpoints = new ExpressEndpoints(app);
  const endpoint0 = endpoints.items.at(0);

  it('print function should output correctly', () => {
    const prefix = 'app';
    const msg0 =
      prefix +
      ' '.repeat(3) +
      expectedEndpoint0.type.padEnd(15) +
      expectedEndpoint0.method.padEnd(12) +
      expectedEndpoint0.route;
    const msg1 = expectedEndpoint0.method.padEnd(12) + expectedEndpoint0.route;

    endpoints.print({ prefix });
    endpoints.print({ color: true });
    endpoints.print({ color: false, prefix, short: true });

    expect(logSpy).toHaveBeenCalled();
    expect(logSpy).toHaveBeenCalledTimes(3);
    expect(logSpy).toHaveBeenCalledWith(msg0);
    expect(logSpy).toHaveBeenCalledWith(msg1);

    logSpy.mockRestore();
    logSpy.mockClear();
  });

  it('items should return list with 1 item(endpoint)', () => {
    expect(endpoints.items).toHaveLength(1);
    expect(endpoint0).not.toBeUndefined();
    expect(endpoint0).not.toBeNull();
    expect(endpoint0).toMatchObject<Endpoint>(expectedEndpoint0);
    expect(endpoint0).toHaveProperty('method', expectedEndpoint0.method);
    expect(endpoint0).toHaveProperty('type', expectedEndpoint0.type);
    expect(endpoint0).toHaveProperty('route', expectedEndpoint0.route);
  });
});

describe('given app has multiple middleware routes and multiple router routes', () => {
  const baseRoute = '/test';
  const routerRoute = '/routerRoute';
  const expectedEndpoints: Endpoint[] = [
    {
      method: 'POST',
      type: 'middleware',
      route: '/^test*$',
    },
    {
      method: 'DELETE',
      type: 'router',
      route: `${baseRoute}${routerRoute}`,
    },
    {
      method: 'GET',
      type: 'router',
      route: `${baseRoute}/:id`,
    },
    {
      method: 'GET',
      type: 'middleware',
      route: `/`,
    },
    {
      method: 'PUT',
      type: 'middleware',
      route: `/:id`,
    },
    {
      method: 'ALL',
      type: 'middleware',
      route: `/*`,
    },
  ];
  const app = express();
  const router = Router();
  const mockRouteFn = jest.fn();
  const endpoint0 = expectedEndpoints[0];
  app.post(endpoint0.route, mockRouteFn);
  router.delete(routerRoute, mockRouteFn);
  router.get(':id', mockRouteFn);
  app.use(baseRoute, router);
  app.get('', mockRouteFn);
  app.put('/:id', mockRouteFn);
  app.all('*', mockRouteFn);
  const endpoints = new ExpressEndpoints(app);

  it('items should return list with 3 items(endpoints)', () => {
    expect(endpoints.items).toHaveLength(expectedEndpoints.length);
    endpoints.items.forEach((endpoint) => {
      expect(endpoint).not.toBeNull();
      expect(endpoint).not.toBeUndefined();
    });
  });

  it('items should have correct endpoint info', () => {
    for (let i = 0; i < endpoints.items.length; i++) {
      const endpoint = expectedEndpoints[i];
      expect(endpoints.items[i]).toMatchObject<Endpoint>(endpoint);
      expect(endpoints.items[i]).toHaveProperty('method', endpoint.method);
      expect(endpoints.items[i]).toHaveProperty('type', endpoint.type);
      expect(endpoints.items[i]).toHaveProperty('route', endpoint.route);
    }
  });
});


src/lib/utils.ts
/**
 * Split a route path into an array of strings
 * @param thing - route path to split
 * @returns array of strings
 * @example
 * utils.split('/test/:id');
 */
const split = (thing: string | RegExp | any): string[] => {
  if (typeof thing === 'string') {
    return thing.split('/');
  } else if (thing.fast_slash) {
    return [''];
  } else if (thing instanceof RegExp) {
    const match = thing
      .toString()
      .replace('\\/?', '')
      .replace('(?=\\/|$)', '$')
      // eslint-disable-next-line no-useless-escape
      .match(/^\/\^((?:\\[.*+?^${}()|[\]\\\/]|[^.*+?^${}()|[\]\\\/])*)\$\//);
    return match != null && match[1]
      ? match[1].replace(/\\(.)/g, '$1').split('/')
      : ['<complex:' + thing.toString() + '>'];
  } else {
    return [''];
  }
};

/**
 * Stringified object of all express api methods to be used for equality check if a route is an `all` middleware,
 * so it can be printed as one line with `ALL` as method instead of multiple lines including all individual methods
 */
const allMethods = JSON.stringify({
  acl: true,
  bind: true,
  checkout: true,
  connect: true,
  copy: true,
  delete: true,
  get: true,
  head: true,
  link: true,
  lock: true,
  'm-search': true,
  merge: true,
  mkactivity: true,
  mkcalendar: true,
  mkcol: true,
  move: true,
  notify: true,
  options: true,
  patch: true,
  post: true,
  propfind: true,
  proppatch: true,
  purge: true,
  put: true,
  rebind: true,
  report: true,
  search: true,
  source: true,
  subscribe: true,
  trace: true,
  unbind: true,
  unlink: true,
  unlock: true,
  unsubscribe: true,
});

export { allMethods, split };


src/lib/interface.ts
/**
 * Type of the endpoint
 * @enum {string} `middleware` | `router`
 */
export type EndpointType = 'middleware' | 'router';

/**
 * Endpoint Interface
 * @example
 * ```js
 * {
 * type: 'middleware',
 * method: 'GET',
 * route: '/testMiddleware'
 * }
 *
 * {
 * type: 'router',
 * method: 'POST',
 * route: '/testRoute'
 * }
 * ```
 */
export interface Endpoint {
  /**
   * Type of the endpoint (`middleware` | `router`)
   * @type {EndpointType}
   */
  type: EndpointType;

  /**
   * HTTP method of the endpoint (can be `ALL` for all methods)
   * @type {string}
   * @example 'GET', 'POST', 'PUT', ..., 'ALL'
   */
  method: string;

  /**
   * Route path of the endpoint (can be a path with params)
   * @type {string}
   * @example '/test' , '/test/:id'
   */
  route: string;
}

/**
 * `ExpressEndpoints` `print` method options
 * @default
 * ```js
 * {
 * prefix: '[ENDPOINT]',
 * color: false,
 * short: false
 * }
 * ```
 * @example
 * ```js
 * {
 * prefix: 'My-Api',
 * color: true,
 * short: false
 * }
 * ```
 */
export interface PrintProps {
  /**
   * Prefix of the message to be printed in the console (only if `short` is false or not provided)
   * @type {string}
   * @default '[ENDPOINT]'
   * @example 'My-Api'
   */
  prefix?: string;

  /**
   * If true, the message will be printed in the console with colors
   * @type {boolean}
   * @default undefined (false)
   */
  color?: boolean;

  /**
   * If true, the message will be printed in the console without prefix and type
   * @type {boolean}
   * @default undefined (false)
   */
  short?: boolean;
}


src/lib/express-endpoint.ts
import type { Endpoint, EndpointType, PrintProps } from '@lib/interface.js';
import { allMethods, split } from '@lib/utils.js';
import chalk from 'chalk';
import type { Application, Express } from 'express';

export class ExpressEndpoints {
  /**
   * List of all endpoints
   * @example
   * const endpoints = new ExpressEndpoints(app);
   * console.log(endpoints.items);
   */
  public items: Endpoint[] = [];

  /**
   * ExpressEndpoints class constructor to extract all endpoints from an express app
   *
   * **NOTE:** create express app and add all routes before creating an instance of ExpressEndpoints
   *
   * @param app - express or Application instance
   * @see {@link https://github.com/igwtcode/express-endpoints#readme more examples and usage guide}
   * @example
   * ```ts
   * import express, { Router } from 'express';
   * // or const express = require('express');
   * import { ExpressEndpoints } from '@igwtcode/express-endpoints';
   * // or const { ExpressEndpoints } = require('@igwtcode/express-endpoints');
   *
   * const app = express();
   * const router = Router();
   * // or const router = express.Router();
   *
   * router.get('/single', (req, res) => res.send(`${req.method} ${req.path}`));
   * router.post('', (req, res) => res.send(`${req.method} ${req.path}`));
   * router.put('/:id', (req, res) => res.send(`${req.method} ${req.path} ${req.params.id}`));
   * router.delete('/:id', (req, res) => res.send(`${req.method} ${req.path} ${req.params.id}`));
   *
   * app.use('/tags', router);
   *
   * app.get('/testGet', (req, res) => res.send(`${req.method} ${req.path}`));
   * app.post('/testPost', (req, res) => res.send(`${req.method} ${req.path}`));
   *
   * app.use((err, req, res, next) => res.status(500).send('error'));
   *
   * app.all('*', (req, res, next) => res.sendStatus(404));
   *
   * const endpoints = new ExpressEndpoints(app); // app is an instance of express or Application
   *
   * console.log('*** with default options');
   * endpoints.print();
   *
   * console.log('\n*** with color and custom prefix');
   * endpoints.print({ color: true, prefix: 'My-Api' });
   *
   * console.log('\n*** colorized and short');
   * endpoints.print({ color: true, short: true });
   *
   * console.log('\n*** items list', endpoints.items);
   *
   * app.listen(3000, () => console.log('listening on port 3000'));
   * ```
   */
  constructor(app: Application | Express) {
    this.init(app);
  }

  /**
   * Initialize the ExpressEndpoints class by extracting all endpoints from the express app and adding them to the items array
   * @param app - express or Application instance
   * @returns void
   */
  private init = (app: Application | Express): void => {
    if (!(app._router && app._router.stack && Array.isArray(app._router.stack))) return;

    for (const layer of app._router.stack) {
      try {
        if (
          layer.name === 'router' &&
          layer.handle &&
          layer.handle.stack &&
          Array.isArray(layer.handle.stack)
        ) {
          for (const stack of layer.handle.stack) {
            if (
              !(
                stack &&
                stack.route &&
                stack.route.methods &&
                typeof stack.route.methods === 'object'
              )
            )
              continue;
            Object.keys(stack.route.methods).forEach((method) => {
              this.addItem('router', method, split(layer.regexp), stack.route.path);
            });
          }
        } else if (
          layer.route &&
          layer.route.methods &&
          typeof layer.route.methods === 'object'
        ) {
          if (JSON.stringify(layer.route.methods) === allMethods) {
            this.addItem('middleware', 'all', [], layer.route.path);
          } else {
            Object.keys(layer.route.methods).forEach((method) => {
              this.addItem('middleware', method, [], layer.route.path);
            });
          }
        }
      } catch (err) {
        console.error(chalk.red('Error while getting endpoints from express app'), err);
      }
    }
  };

  /**
   * Add an endpoint to the list of endpoints (`items`)
   * @param type - (`middleware` or `router`)
   * @param method - HTTP method (GET, POST, ..., ALL)
   * @param basePath - Base path of the endpoint (e.g. `/tags`)
   * @param route - Route of the endpoint (e.g. `/single`)
   * @returns {void}
   */
  private addItem = (
    type: EndpointType,
    method: string,
    basePath: string | string[],
    route: string,
  ): void => {
    const bp = Array.isArray(basePath) ? basePath : [basePath];
    this.items.push({
      type,
      method: method.toUpperCase(),
      route:
        bp.join('/') +
          (route.startsWith('/') ? route : route.length ? `/${route}` : '') || '/',
    });
  };

  /**
   * Get message to print in the console
   * @param endpoint - Endpoint object to print
   * @param props - options (prefix, short, color) to customize the message
   * @returns {string} - message to print in the console
   */
  private getMessage = (endpoint: Endpoint, props?: PrintProps): string => {
    const p1 = !props?.short ? (props?.prefix || '[ENDPOINT]') + ' '.repeat(3) : '';
    const p2 = !props?.short ? endpoint.type.padEnd(15) : '';
    const p3 = endpoint.method.padEnd(12);
    const p4 = endpoint.route;
    return props?.color
      ? chalk.green(p1) + chalk.cyan(p2) + chalk.blue(p3) + chalk.blueBright(p4)
      : p1 + p2 + p3 + p4;
  };

  /**
   * Print all endpoints in the console with the specified properties
   * or the default values if no properties are specified
   * @param props - options (prefix, short, color) to customize the message
   * @returns {void}
   * @example
   * const endpoints = new ExpressEndpoints(app);
   * endpoints.print(); // using default values `{ prefix: '[ENDPOINT]', color: false, short: false }`
   * endpoints.print({ color: true });
   * endpoints.print({ color: true, short: true });
   * endpoints.print({ prefix: '[MyApp]' });
   * endpoints.print({ prefix: 'My Api', color: true });
   */
  print = (props?: PrintProps): void => {
    if (this.items.length === 0) return;
    this.items.forEach((endpoint: Endpoint) => {
      console.log(this.getMessage(endpoint, props));
    });
  };
}
